src/sparse/prod.rs

```diff
@@ -10,19 +10,49 @@ use num_traits::Num;
 /// Compute the dot product of two sparse vectors, using binary search to find matching indices.
 ///
 /// Runs in O(MlogN) time, where M and N are the number of non-zero entries in each vector.
-pub fn csvec_dot_by_binary_search<N, I>(
-    vec1: CsVecViewI<N, I>,
-    vec2: CsVecViewI<N, I>,
+pub fn csvec_dot_by_binary_search<N, I, A, B>(
+    vec1: CsVecViewI<A, I>,
+    vec2: CsVecViewI<B, I>,
 ) -> N
 where
     I: SpIndex,
-    N: crate::MulAcc + num_traits::Zero,
+    N: crate::MulAcc<A, B> + num_traits::Zero,
 {
-    let (mut idx1, mut val1, mut idx2, mut val2) = if vec1.nnz() < vec2.nnz() {
-        (vec1.indices(), vec1.data(), vec2.indices(), vec2.data())
+    //Check vec1.nnz<vec2.nnz ; call self with args flipped if not
+    // can't work with the original method with seperate A,B types
+    // because the val1 / val2 would be swapped from &[A]<->&[B]
+    // lambda passed into the subroutine dodges the need for MulAcc<B,A>
+    // we might want to fit non-commutative plugin MulAcc functions, eg 'vectors of small dense matrices' etc.
```

Comment:
Maybe better phrased as:
Optimisation: Reverse the dot product `vec1` and `vec2`, but preserve possibly non-commutative `MulAcc` through a lamba.

Reply:
Would be interesting to see if the performance is improved by using `NnzOrZip` in this case, so we don't have to binary search one array all the time.

Reply:
right all these style fixes are fine by me.  foo_impl is a better name for 'foo_sub(routine)'  _sub in a maths lib is indeed confusing (i was also tempted to make it a local lambda but its too much nesting)

Are you integrating these tweaks directly or do you want me to replicate them prior to the pull request being approved?

Reply:
I can fix them this afternoon, but you may also fix them if you prefer